// json-http-proxy/lib/upstream.js
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

Object.defineProperty(exports, "__esModule", {
	value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var url = require('url');

var Upstream = exports.Upstream = (function () {
	function Upstream(name) {
		var opts = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

		_classCallCheck(this, Upstream);

		this.name = name;
		this.protocol = opts.protocol || 'http:';
		this.hostname = opts.hostname || 'localhost';
		this.port = opts.port || 80;
		this.path = opts.path || '/';
		this.proxy = opts.proxy;
	}

	_createClass(Upstream, [{
		key: 'handle',
		value: function handle(req, res) {
			this.proxy.web(req, res, {
				target: url.format(this)
			});
		}
	}]);

	return Upstream;
})();
