// @boundlessgeo/sdk/components/FilterHelp.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp; /*
                    * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
                    * Licensed under the Apache License, Version 2.0 (the "License").
                    * You may not use this file except in compliance with the License.
                    * You may obtain a copy of the License at
                    * http://www.apache.org/licenses/LICENSE-2.0
                    * Unless required by applicable law or agreed to in writing, software
                    * distributed under the License is distributed on an "AS IS" BASIS,
                    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                    * See the License for the specific language governing permissions and limitations under the License.
                    */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _Popover = require('material-ui/Popover');

var _Popover2 = _interopRequireDefault(_Popover);

var _helpOutline = require('material-ui/svg-icons/action/help-outline');

var _helpOutline2 = _interopRequireDefault(_helpOutline);

var _reactIntl = require('react-intl');

var _Paper = require('material-ui/Paper');

var _Paper2 = _interopRequireDefault(_Paper);

require('./FilterHelp.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var messages = (0, _reactIntl.defineMessages)({
  introtextprefix: {
    'id': 'filterhelp.introtextprefix',
    'defaultMessage': 'Type in'
  },
  introtextsuffix: {
    'id': 'filterhelp.introtextsuffix',
    'defaultMessage': 'a filter expression to narrow your search to one or more attributes.'
  },
  exampletext: {
    'id': 'filterhelp.exampletext',
    'defaultMessage': 'Examples:'
  },
  example1filter: {
    'id': 'filterhelp.example1filter',
    'defaultMessage': 'population > 100000'
  },
  example1description: {
    'id': 'filterhelp.example1description',
    'defaultMessage': 'All features with a population greater than 100000'
  },
  example2filter: {
    'id': 'filterhelp.example2filter',
    'defaultMessage': 'sovereignt == "United Kingdom" and not (subregion like "Europe")'
  },
  example2description: {
    'id': 'filterhelp.example2description',
    'defaultMessage': 'All british colonies not in Europe'
  },
  generalsection: {
    'id': 'filterhelp.generalsection',
    'defaultMessage': 'Expressions may contain attribute names, strings, numbers, and operators. Attribute names may be surrounded  by single quotes (\') to reduce ambiguity. Strings must be surrounded by double quotes ("). Numbers and operators are unquoted.'
  },
  operatortext: {
    'id': 'filterhelp.operatortext',
    'defaultMessage': 'Supported operators are:'
  },
  operator1: {
    'id': 'filterhelp.operator1',
    'defaultMessage': 'a == a'
  },
  operator1description: {
    'id': 'filterhelp.operator1description',
    'defaultMessage': 'a exactly equals b'
  },
  operator2: {
    'id': 'filterhelp.operator2',
    'defaultMessage': 'a != a'
  },
  operator2description: {
    'id': 'filterhelp.operator2description',
    'defaultMessage': 'a does not equal b'
  },
  operator3: {
    'id': 'filterhelp.operator3',
    'defaultMessage': 'a < a'
  },
  operator3description: {
    'id': 'filterhelp.operator3description',
    'defaultMessage': 'a is less than b'
  },
  operator4: {
    'id': 'filterhelp.operator4',
    'defaultMessage': 'a <= b'
  },
  operator4description: {
    'id': 'filterhelp.operator4description',
    'defaultMessage': 'a is less than or equal to b'
  },
  operator5: {
    'id': 'filterhelp.operator5',
    'defaultMessage': 'a > b'
  },
  operator5description: {
    'id': 'filterhelp.operator5description',
    'defaultMessage': 'a is greater than b'
  },
  operator6: {
    'id': 'filterhelp.operator6',
    'defaultMessage': 'a >= b'
  },
  operator6description: {
    'id': 'filterhelp.operator6description',
    'defaultMessage': 'a is greater than or equal to b'
  },
  operator7: {
    'id': 'filterhelp.operator7',
    'defaultMessage': 'a like b'
  },
  operator7description: {
    'id': 'filterhelp.operator7description',
    'defaultMessage': 'b contains a (case insensitive)'
  },
  operator8: {
    'id': 'filterhelp.operator8',
    'defaultMessage': 'a in (x,y,z)'
  },
  operator8description: {
    'id': 'filterhelp.operator8description',
    'defaultMessage': 'a is equal to one or more values in the list x,y,z'
  },
  operator9: {
    'id': 'filterhelp.operator9',
    'defaultMessage': 'e and f'
  },
  operator9description: {
    'id': 'filterhelp.operator9description',
    'defaultMessage': 'Matches both expressions e and f'
  },
  operator10: {
    'id': 'filterhelp.operator10',
    'defaultMessage': 'e or f'
  },
  operator10description: {
    'id': 'filterhelp.operator10description',
    'defaultMessage': 'Matches either expression e or f'
  },
  operator11: {
    'id': 'filterhelp.operator11',
    'defaultMessage': 'not e'
  },
  operator11description: {
    'id': 'filterhelp.operator11description',
    'defaultMessage': 'Returns all results that do not match the expression e'
  },
  parantheses: {
    'id': 'filterhelp.parantheses',
    'defaultMessage': 'Any expression may be surrounded by parentheses for clarity (e.g. when combining ands and ors).'
  }
});

/**
 * Help dialog for filter fields.
 *
 * ```xml
 * <FilterHelp />
 * ```
 */
var FilterHelp = (_temp = _class = function (_React$Component) {
  _inherits(FilterHelp, _React$Component);

  function FilterHelp(props) {
    _classCallCheck(this, FilterHelp);

    var _this = _possibleConstructorReturn(this, (FilterHelp.__proto__ || Object.getPrototypeOf(FilterHelp)).call(this, props));

    _this.state = {
      help: false
    };
    return _this;
  }

  _createClass(FilterHelp, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.helpElement = _reactDom2.default.findDOMNode(this.refs.help);
    }
  }, {
    key: '_onToggleHelp',
    value: function _onToggleHelp() {
      this.setState({ help: !this.state.help });
    }
  }, {
    key: 'render',
    value: function render() {
      var formatMessage = this.props.intl.formatMessage;

      var introText = formatMessage(messages.introtextprefix) + ' ';
      introText += formatMessage(messages.introtextsuffix);
      return _react2.default.createElement(
        'span',
        { className: 'filter-help' },
        _react2.default.createElement(_helpOutline2.default, { ref: 'help', onClick: this._onToggleHelp.bind(this) }),
        _react2.default.createElement(
          _Popover2.default,
          { style: this.props.style, open: this.state.help, onRequestClose: this._onToggleHelp.bind(this), anchorEl: this.helpElement, anchorOrigin: { 'horizontal': 'left' } },
          _react2.default.createElement(
            'p',
            null,
            introText
          ),
          _react2.default.createElement(
            'p',
            null,
            formatMessage(messages.exampletext)
          ),
          _react2.default.createElement(
            'ul',
            null,
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { style: this.props.monoStyle, zDepth: 0 },
                formatMessage(messages.example1filter)
              ),
              _react2.default.createElement('br', null),
              formatMessage(messages.example1description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { style: this.props.monoStyle, zDepth: 0 },
                formatMessage(messages.example2filter)
              ),
              _react2.default.createElement('br', null),
              formatMessage(messages.example2description)
            )
          ),
          _react2.default.createElement(
            'p',
            null,
            formatMessage(messages.generalsection)
          ),
          _react2.default.createElement(
            'p',
            null,
            formatMessage(messages.operatortext)
          ),
          _react2.default.createElement(
            'ul',
            null,
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator1)
              ),
              formatMessage(messages.operator1description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator2)
              ),
              formatMessage(messages.operator2description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator3)
              ),
              formatMessage(messages.operator3description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator4)
              ),
              formatMessage(messages.operator4description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator5)
              ),
              formatMessage(messages.operator5description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator6)
              ),
              formatMessage(messages.operator6description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator7)
              ),
              formatMessage(messages.operator7description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator8)
              ),
              formatMessage(messages.operator8description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator9)
              ),
              formatMessage(messages.operator9description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator10)
              ),
              formatMessage(messages.operator10description)
            ),
            _react2.default.createElement(
              'li',
              null,
              _react2.default.createElement(
                _Paper2.default,
                { zDepth: 0, style: this.props.monoStyle },
                formatMessage(messages.operator11)
              ),
              formatMessage(messages.operator11description)
            )
          ),
          _react2.default.createElement(
            'p',
            null,
            formatMessage(messages.parantheses)
          )
        )
      );
    }
  }]);

  return FilterHelp;
}(_react2.default.Component), _class.propTypes = {
  /**
   * Style config.
   */
  style: _react2.default.PropTypes.object,
  /**
   * Mono style config.
   */
  monoStyle: _react2.default.PropTypes.object,
  /**
   * @ignore
   */
  intl: _reactIntl.intlShape.isRequired
}, _class.defaultProps = {
  style: {
    padding: '12px',
    margin: '36px',
    maxWidth: '600px'
  },
  monoStyle: {
    fontWeight: 'bold',
    display: 'inline-block',
    minWidth: '80px'
  }
}, _temp);
exports.default = (0, _reactIntl.injectIntl)(FilterHelp);
