// @boundlessgeo/sdk/components/Header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp; /*
                    * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
                    * Licensed under the Apache License, Version 2.0 (the "License").
                    * You may not use this file except in compliance with the License.
                    * You may obtain a copy of the License at
                    * http://www.apache.org/licenses/LICENSE-2.0
                    * Unless required by applicable law or agreed to in writing, software
                    * distributed under the License is distributed on an "AS IS" BASIS,
                    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                    * See the License for the specific language governing permissions and limitations under the License.
                    */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _AppBar = require('material-ui/AppBar');

var _AppBar2 = _interopRequireDefault(_AppBar);

var _Logo = require('./Logo');

var _Logo2 = _interopRequireDefault(_Logo);

require('./Header.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * A header toolbar for holding buttons
 *
 * ```xml
 * <Header  />
 * ```
 */

var Header = (_temp = _class = function (_React$Component) {
  _inherits(Header, _React$Component);

  function Header(props) {
    _classCallCheck(this, Header);

    var _this = _possibleConstructorReturn(this, (Header.__proto__ || Object.getPrototypeOf(Header)).call(this, props));

    _this.state = {
      menuOpen: false
    };

    _this.handleMenuTouchTap = function (event) {
      // This prevents ghost click.
      event.preventDefault();
      _this.setState({
        menuOpen: true,
        anchorEl: event.currentTarget
      });
    };

    _this.handleMenuRequestClose = function () {
      _this.setState({
        menuOpen: false
      });
    };

    return _this;
  }

  _createClass(Header, [{
    key: 'render',
    value: function render() {
      var title;
      if (this.props.logo) {
        title = _react2.default.createElement(_Logo2.default, { src: this.props.logo, onTouchTap: this.props.onLeftIconTouchTap, visible: this.props.showLeftIcon, tooltip: this.props.title });
      } else {
        title = this.props.title;
      }

      return _react2.default.createElement(_AppBar2.default, {
        style: this.props.style,
        showMenuIconButton: this.props.showLeftIcon,
        title: title,
        iconElementRight: _react2.default.createElement(
          'div',
          null,
          this.props.children
        ),
        onLeftIconButtonTouchTap: this.props.onLeftIconTouchTap
      });
    }
  }]);

  return Header;
}(_react2.default.Component), _class.propTypes = {
  /**
   * Title of the Header
   */
  title: _react2.default.PropTypes.string,
  /**
   * Style config.
   */
  style: _react2.default.PropTypes.object,
  /**
  * Array of menu items to add to the header
  */
  leftMenuItems: _react2.default.PropTypes.node,
  /**
   * Image file to be used for logo
   */
  logo: _react2.default.PropTypes.string,
  /**
  * @ignore
  * Callback for left Icon
  **/
  onLeftIconTouchTap: _react2.default.PropTypes.func,
  /**
   * Determines if the left icon will display next to the title.
   */
  showLeftIcon: _react2.default.PropTypes.bool,
  /**
   * Child nodes
   */
  children: _react2.default.PropTypes.node
}, _class.childContextTypes = {
  muiTheme: _react2.default.PropTypes.object.isRequired
}, _class.defaultProps = {
  showLeftIcon: true
}, _temp);
exports.default = Header;
