// @boundlessgeo/sdk/actions/GeocodingActions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _GeocodingConstants = require('../constants/GeocodingConstants');

var _GeocodingConstants2 = _interopRequireDefault(_GeocodingConstants);

var _AppDispatcher = require('../dispatchers/AppDispatcher');

var _AppDispatcher2 = _interopRequireDefault(_AppDispatcher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

exports.default = {
  showSearchResult: function showSearchResult(results) {
    _AppDispatcher2.default.handleAction({
      type: _GeocodingConstants2.default.SHOW_SEARCH_RESULTS,
      searchResults: results
    });
  },
  clearSearchResult: function clearSearchResult() {
    _AppDispatcher2.default.handleAction({
      type: _GeocodingConstants2.default.CLEAR_SEARCH_RESULT
    });
  },
  zoomToResult: function zoomToResult(result) {
    _AppDispatcher2.default.handleAction({
      type: _GeocodingConstants2.default.ZOOM_TO_RESULT,
      result: result
    });
  }
};
