// @boundlessgeo/sdk/actions/LayerActions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _LayerConstants = require('../constants/LayerConstants');

var _LayerConstants2 = _interopRequireDefault(_LayerConstants);

var _AppDispatcher = require('../dispatchers/AppDispatcher');

var _AppDispatcher2 = _interopRequireDefault(_AppDispatcher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

exports.default = {
  removeLayer: function removeLayer(layer, group) {
    _AppDispatcher2.default.handleAction({
      type: _LayerConstants2.default.REMOVE_LAYER,
      layer: layer,
      group: group
    });
  },
  moveLayer: function moveLayer(dragIndex, hoverIndex, layer, group) {
    _AppDispatcher2.default.handleAction({
      type: _LayerConstants2.default.MOVE_LAYER,
      dragIndex: dragIndex,
      hoverIndex: hoverIndex,
      layer: layer,
      group: group
    });
  },
  editLayer: function editLayer(layer) {
    _AppDispatcher2.default.handleAction({
      type: _LayerConstants2.default.EDIT_LAYER,
      layer: layer
    });
  },
  styleLayer: function styleLayer(layer) {
    _AppDispatcher2.default.handleAction({
      type: _LayerConstants2.default.STYLE_LAYER,
      layer: layer
    });
  }
};
