// @boundlessgeo/sdk/actions/SelectActions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _SelectConstants = require('../constants/SelectConstants');

var _SelectConstants2 = _interopRequireDefault(_SelectConstants);

var _AppDispatcher = require('../dispatchers/AppDispatcher');

var _AppDispatcher2 = _interopRequireDefault(_AppDispatcher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

exports.default = {
  toggleFeature: function toggleFeature(layer, feature) {
    _AppDispatcher2.default.handleAction({
      type: _SelectConstants2.default.TOGGLE_FEATURE,
      layer: layer,
      feature: feature
    });
  },
  clear: function clear(layer, filter) {
    _AppDispatcher2.default.handleAction({
      type: _SelectConstants2.default.CLEAR,
      layer: layer,
      filter: filter
    });
  },
  selectFeatures: function selectFeatures(layer, features, clear) {
    _AppDispatcher2.default.handleAction({
      type: _SelectConstants2.default.SELECT_FEATURES,
      layer: layer,
      features: features,
      clear: clear
    });
  }
};
