// @boundlessgeo/sdk/actions/ToolActions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _ToolConstants = require('../constants/ToolConstants');

var _ToolConstants2 = _interopRequireDefault(_ToolConstants);

var _AppDispatcher = require('../dispatchers/AppDispatcher');

var _AppDispatcher2 = _interopRequireDefault(_AppDispatcher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

exports.default = {
  showPopup: function showPopup(feature, layer) {
    _AppDispatcher2.default.handleAction({
      type: _ToolConstants2.default.SHOW_POPUP,
      feature: feature,
      layer: layer
    });
  },
  showEditPopup: function showEditPopup(feature, layer, callback) {
    _AppDispatcher2.default.handleAction({
      type: _ToolConstants2.default.SHOW_EDIT_POPUP,
      feature: feature,
      layer: layer,
      callback: callback
    });
  },
  startPlayback: function startPlayback() {
    _AppDispatcher2.default.handleAction({
      type: _ToolConstants2.default.START_PLAYBACK
    });
  },
  stopPlayback: function stopPlayback() {
    _AppDispatcher2.default.handleAction({
      type: _ToolConstants2.default.STOP_PLAYBACK
    });
  },
  disableAllTools: function disableAllTools() {
    _AppDispatcher2.default.handleAction({
      type: _ToolConstants2.default.DISABLE_ALL_TOOLS
    });
  },
  enableAllTools: function enableAllTools() {
    _AppDispatcher2.default.handleAction({
      type: _ToolConstants2.default.ENABLE_ALL_TOOLS
    });
  },
  activateTool: function activateTool(tool, toggleGroup, toolId) {
    _AppDispatcher2.default.handleAction({
      type: _ToolConstants2.default.ACTIVATE_TOOL,
      tool: tool,
      toggleGroup: toggleGroup,
      toolId: toolId
    });
  }
};
